### ISSUE TITLE
Usage of Multinomial with logits

### ISSUE THREAD

**Issue by karanchahal** [openai/spinningup#97]
Hello, 

I was looking at network architecture for the VPG code. Upon receiving the logits I saw the line:

```
pi = tf.squeeze(tf.multinomial(logits,1), axis=1)
```

According to me, this samples any random index out the log probabilities output of the network across the various actions. However, shouldn't we choose the action with the highest log probability as that means the network is most confident in that action ?

Multinomial takes out a random index out of a probability distribution right ?

**Comment by colllin** [comment]
@karanchahal — I think this is the key difference between a _stochastic_ policy and a _deterministic_ policy.  A stochastic policy [models a distribution over actions, and draws an action according to this distribution](https://www.reddit.com/r/MachineLearning/comments/4l4yea/stochastic_vs_deterministic_policy_gradients_what/).  After reading your question, I wondered if during "test" time it would be better to choose the action with the highest probability, and [this discussion has some good insight](https://www.reddit.com/r/reinforcementlearning/comments/8zqvbh/d_policy_gradient_testtime_action_selection/) there.

**Comment by karanchahal** [comment]
Thank you for your answer ! Aah what you said makes sense but I am a little confused. What does "drawing an action according to a distribution " exactly mean ? For a multinomial distribution, is drawing an action biased towards actions with higher values?

**Comment by colllin** [comment]
I'm not an expert, but check out the docs for [tf.random.multinomial](https://www.tensorflow.org/api_docs/python/tf/random/multinomial).  It looks like the first argument, `logits`, is a 2-D tensor (batch x probabilities), where each row `[i, :] represents the unnormalized log-probabilities for all classes`.  So, yes, higher values (representing a higher log-probability) would cause those classes to be drawn more frequently.

**Comment by karanchahal** [comment]
Yep thank you ! Understood

**Comment by wiljohnhong** [comment]
Hello, I have a another question about this line of code
```
pi = tf.squeeze(tf.multinomial(logits,1), axis=1)
```
Shouldn't it be `pi = tf.squeeze(tf.multinomial(logp_all,1), axis=1)` instead? Since  `tf.random.multinomial` should receive `logits` which represent the log-probability.
